src/wgetopt.rs

```diff
@@ -25,341 +26,337 @@ Cambridge, MA 02139, USA.  */
+                    left += len;
                 }
-                // Exclude the moved top segment from further swapping.
-                bottom += len;
             }
         }
 
-        // Update records for the slots the non-options now occupy.
-        self.first_nonopt += self.woptind - self.last_nonopt;
-        self.last_nonopt = self.woptind;
+        // Update the indices to indicate the new positions of the non-option
+        // arguments.
+        self.first_nonopt += self.wopt_index - self.last_nonopt;
+        self.last_nonopt = self.wopt_index;
     }
 
     /// Initialize the internal data when the first call is made.
-    fn _wgetopt_initialize(&mut self) {
-        // Start processing options with ARGV-element 1 (since ARGV-element 0 is the program name); the
-        // sequence of previously skipped non-option ARGV-elements is empty.
+    fn initialize(&mut self) {
+        // Skip the first element since it's just the program name.
         self.first_nonopt = 1;
         self.last_nonopt = 1;
-        self.woptind = 1;
-        self.nextchar = empty_wstr();
+        self.wopt_index = 1;
+        self.remaining_text = empty_wstr();
 
         let mut optstring = self.shortopts;
 
-        // Determine how to handle the ordering of options and nonoptions.
+        // Set ordering and strip markers.
         if optstring.char_at(0) == '-' {
-            self.ordering = Ordering::RETURN_IN_ORDER;
+            self.ordering = Ordering::ReturnInOrder;
             optstring = &optstring[1..];
         } else if optstring.char_at(0) == '+' {
-            self.ordering = Ordering::REQUIRE_ORDER;
+            self.ordering = Ordering::RequireOrder;
             optstring = &optstring[1..];
         } else {
-            self.ordering = Ordering::PERMUTE;
+            self.ordering = Ordering::Permute;
         }
 
         if optstring.char_at(0) == ':' {
-            self.missing_arg_return_colon = true;
+            self.return_colon = true;
             optstring = &optstring[1..];
         }
 
         self.shortopts = optstring;
         self.initialized = true;
     }
 
-    /// Advance to the next ARGV-element.
-    /// \return Some(\0) on success, or None or another value if we should stop.
-    fn _advance_to_next_argv(&mut self) -> Option<char> {
-        let argc = self.argc();
-        if self.ordering == Ordering::PERMUTE {
-            // If we have just processed some options following some non-options, exchange them so
-            // that the options come first.
-            if self.first_nonopt != self.last_nonopt && self.last_nonopt != self.woptind {
-                self.exchange();
-            } else if self.last_nonopt != self.woptind {
-                self.first_nonopt = self.woptind;
+    /// Advance to the next element in `argv`. Returns nothing on success, or
+    /// an optional value if we need to stop.
+    fn next_argv(&mut self) -> Result<(), Option<char>> {
```

Comment:
This return type doesn't match the comment.

Before this change, we had the following return types:

1. `None` => options are fully processed, caller should exit their loop, `woptind` now points at remaining arguments.
2. `Some(c) if c != '\0'` => a special return; I think `NON_OPTION_CHAR` is the only possible return here
3. `Some('\0')` => an option was encountered and the caller should handle it. It can be determined by inspecting `self.argv[self.woptind]` and `nextchar`

Probably worth pulling out this return type into a real enum; in particular `Result` suggests an error when it's actually fine and expected.